File: leetcode/0697-degree-of-an-array/solution.py
from typing import List


class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        first_index = {}
        last_index = {}
        count = {}

        for i, x in enumerate(nums):
            if x not in first_index:
                first_index[x] = i
            last_index[x] = i
            count[x] = count.get(x, 0) + 1

        degree = max(count.values())
        answer = len(nums)
        for x, freq in count.items():
            if freq == degree:
                answer = min(answer, last_index[x] - first_index[x] + 1)

        return answer

File: leetcode/0697-degree-of-an-array/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        unordered_map<int, int> firstIndex, lastIndex, count;

        for (int i = 0; i < (int) nums.size(); i++) {
            int x = nums[i];
            if (firstIndex.find(x) == firstIndex.end()) {
                firstIndex[x] = i;
            }
            lastIndex[x] = i;
            count[x]++;
        }

        int degree = 0;
        for (auto& [x, freq] : count) {
            degree = max(degree, freq);
        }

        int answer = (int) nums.size();
        for (auto& [x, freq] : count) {
            if (freq == degree) {
                answer = min(answer, lastIndex[x] - firstIndex[x] + 1);
            }
        }

        return answer;
    }
};
